packages/types/src/action.ts

```diff
@@ -42,3 +74,11 @@ export type AsyncAction<Pkg extends Package, Input = null> = [Input] extends [
       actions: ResolveActions<Pkg["actions"]>;
       libraries: Pkg["libraries"];
     }) => (args: Input) => Promise<void>;
+
+type Context = Koa.ParameterizedContext<Koa.DefaultState, Koa.DefaultContext>;
+
```

Comment:
I think this is correct? We have to support every use case like the following, right?

```js
// SYNC
// 1. sync, no params
action: ({state, ...rest }) => {}

// 2. sync, with params
action: ({state, ...rest }) => params =>  {}

// ASYNC
// 3. async, no params
action: async ({state, ...rest }) =>  {}

// 4. async, with params
action: ({state, ...rest }) => async (params) => {}
```

Reply:
oh, I realized that `beforeSSR` is supposed to always be async (at least according to the docs)

Reply:
nevermind, that's not true. It can be both.